app/models/provisioning_template.rb

```diff
@@ -50,11 +50,14 @@ def used_taxonomy_ids(type)
     end
   }
 
-  # TODO: review if we can improve SQL
   def self.template_ids_for(hosts)
-    hosts.with_os.map do |host|
-      host.provisioning_template.try(:id)
-    end.uniq.compact
+    oses = hosts.with_os.uniq.pluck(:operatingsystem_id)
+    hostgroups = hosts.with_os.uniq.pluck(:hostgroup_id) | [nil]
+    environments = hosts.with_os.uniq.pluck(:environment_id) | [nil]
+    templates = ProvisioningTemplate.reorder(nil).joins(:operatingsystems, :template_kind).where('operatingsystems.id' => oses, 'template_kinds.name' => 'provision')
+    ids = templates.joins(:template_combinations).where("template_combinations.hostgroup_id" => hostgroups, "template_combinations.environment_id" => environments).uniq.pluck(:id)
```

Comment:
OK, I think that justifies it. I wonder if we shouldn't display resources regardless they are used or not, but that's not related to this PR...